neonwranglerpy/lib/extract_hsi_to_tif.py

```diff
@@ -0,0 +1,263 @@
+    subArray_cols = clipIndex['xMax'] - clipIndex['xMin']
+
+    stackedArray = np.zeros((subArray_rows, subArray_cols, len(bands)), dtype=np.int16)
+    band_clean_dict = {}
+    band_clean_names = []
+
+    for i in range(len(bands)):
+        band_clean_names.append("b" + str(bands[i]) + "_refl_clean")
+        band_clean_dict[band_clean_names[i]] = subset_clean_band(
+            reflArray, reflArray_metadata, clipIndex, bands[i])
+        stackedArray[..., i] = band_clean_dict[band_clean_names[i]]
+
+    return stackedArray
+
+
+def subset_clean_band(reflArray, reflArray_metadata, clipIndex, bandIndex):
+    bandCleaned = reflArray[clipIndex['yMin']:clipIndex['yMax'],
+                            clipIndex['xMin']:clipIndex['xMax'],
+                            bandIndex - 1].astype(np.int16)
+
+    return bandCleaned
+
+#    array2raster(tilename, hyperspec_raster, sub_meta, clipExtent, save_dir)
+
+def array2raster(newRaster, reflBandArray, reflArray_metadata, extent, ras_dir, invert_axes = True):
+    """
+    newRaster: filename of the raster object
+    reflBandArray: Clipped wavelength data,
+    reflArray_metadata: Clipped wavelength metadata
+    extent: The UTM coordinate extent
+    ras_dir: Where to save the file
+    """
+    print(reflBandArray.shape)
+    if invert_axes is True:
+        cols = reflBandArray.shape[1]
+        rows = reflBandArray.shape[0]
+        bands = reflBandArray.shape[2]
+        reflBandArray = np.moveaxis(reflBandArray,2,0)  
+    else:
+        cols = reflBandArray.shape[1]
+        rows = reflBandArray.shape[2]
+        bands = reflBandArray.shape[0]
+
+    originX = extent['xMin']
+    originY = extent['yMax']
+    res = reflArray_metadata['res']['pixelWidth']
+    transform = Affine.translation(originX, originY) * Affine.scale(res, -res)
+    with rasterio.open(
+        "{}/{}".format(ras_dir,newRaster),
+         'w',
+         driver='GTiff',
+         height=rows,
+         width=cols,
+         count=bands,
+         dtype=reflBandArray.dtype,
+         crs=rasterio.crs.CRS.from_dict(init='epsg:'+str(reflArray_metadata["epsg"])),
+         transform=transform,) as dst:
+        dst.write(reflBandArray)
+    return(reflBandArray)
+        
+
+
+def calc_clip_index(clipExtent, h5Extent, xscale=1, yscale=1):
+    """Extract numpy index for the utm coordinates"""
+    h5rows = h5Extent['yMax'] - h5Extent['yMin']
+    h5cols = h5Extent['xMax'] - h5Extent['xMin']
+
+    ind_ext = {}
+    ind_ext['xMin'] = round((clipExtent['xMin'] - h5Extent['xMin']) / xscale)
+    ind_ext['xMax'] = round((clipExtent['xMax'] - h5Extent['xMin']) / xscale)
+    ind_ext['yMax'] = round(h5rows - (clipExtent['yMin'] - h5Extent['yMin']) / yscale)
+    ind_ext['yMin'] = round(h5rows - (clipExtent['yMax'] - h5Extent['yMin']) / yscale)
+
+    return ind_ext
+
+
+def generate_raster(h5_path, save_dir, rgb_filename=None, bands="no_water", correction="all", bounds = False, suffix=None):
```

Comment:
`generate_raster` is the main function right? , it will work same as [extract_elevation_data](https://github.com/weecology/neonVegWrangleR/blob/c34c1083b18a5aa6f0d9d72d16a51f4f6eaee0bf/R/clip_plot.R#L77) function?

Reply:
yes, the generate_raster function will substitute the extract_elevation_data. The function will automatically save 2 geotiffs, 1 with the reflectance, 1 with the solar/sensor angles that can be used for  calculating the topographic and bdrf corrections

Reply:
`rgb_filename` can't be `None` as [L#254](https://github.com/weecology/neonwranglerpy/blob/28374b6ccb564b067437ea69ec41d156a44f66b6/neonwranglerpy/lib/extract_hsi_to_tif.py#L254) will throw an error as : `TypeError: expected str, bytes or os.PathLike object, not NoneType
`

Reply:
that is intentional because it needs a filename, but definitively I should add a better error handler 👍

Reply:
can you please give me an example of `rgb_filename`